open-company/open-company-auth
# Better handle of bot removal

Card: https://trello.com/c/q7yRcvwe

### Merge after #103 

PRs:
- ~~[Email](https://github.com/open-company/open-company-email/pull/59)~~
- ~~[Storage](https://github.com/open-company/open-company-storage/pull/224)~~
- ~~[Lib](https://github.com/open-company/open-company-lib/pull/75)~~
- ~~[Infra](https://github.com/belucid/open-company-infrastructure/pull/122)~~
- ~~[Web](https://github.com/open-company/open-company-web/pull/1058)~~
- ~~Lib~~

To test (locally):
- setup ngrok and local slack events to your dev.
- onboard with our opencompanyhq Slack team
- create the org (let's call it Org 1)
- add the bot
- set the current user notification settings all to Slack
- now go [here](https://opencompanyhq.slack.com/apps/A5SGHA79P-carrot-local-development)
- scroll to the bottom
- click Remove app
- open notifications panel and wait for data refresh
- [x] all set to email now?
- [ ] do you see a message in #customer-support the org slug linked to its stabejo page and the number of admins?
- [x] did you get an email with a link to the integrations panel for the right org?
- add the Slack bot again
- go to Notifications panel and change all to Slack
- onboard a user via email and create a new email org (let's call it Org 2)
- invite via email the user previously onboarded via Slack to the new team
- accept the invite (or simply relogin with your Slack user)
- go to Notifications panel while on the new
- [x] do you see Slack as selected medium in all 3?
- [x] can you change to email on all? or to in-app for the last 2?
- [x] can you NOT change back to Slack? (this is because you have the Slack bot on another team but not the current one, but the medium settings are per user, not per user per team)
- now go back to Org 1
- open Notifications panel
- set digest to Slack, notifications to email and reminders in-app
- now go [here](https://opencompanyhq.slack.com/apps/A5SGHA79P-carrot-local-development)
- scroll to the bottom
- click Remove app
- [x] Notifications preference are in order: email, email, in-app
- [x] did you get an email?
- [ ] message in #customer-support with 2 orgs this time
- no go to Org 1
- add the bot from integrations panel (2 steps)
- go to Notifications panel
- set digest to email, notifications to in-app and reminders to Slack
- now login to one of our Slack test org
- onboard on Carrot with it and create an org
- add the bot to the Carrot org
- *on Slack* invite yourself to oc-test-org1 (or 2) using the email you have for the user in Org 1 and Org 2
- accept the Slack invite
- now with the first user of oc-test-org1 login to carrot and invite your user via Slack
- with your user login to Carrot (you can use opencompanyhq or oc-test-org1 account both)
- once you are in check Notifications preference are still in order: email, in-app and Slack
- now go [here](https://opencompanyhq.slack.com/apps/A5SGHA79P-carrot-local-development)
- scroll to the bottom
- click Remove app
- with your main user go to Notifications panel
- [x] do you still see in order: email, in-app and Slack? This is because your user is still associated with a team that has a Slack bot and it has also a corresponding Slack user in that team
- add the bot back if you want

Before you merge:
- upgrade lib version in `open-company-lib/project.clj` to lose the alpha
- run `lein deploy release` to release the lib with the new version
- now in `open-company-auth/project.clj` update the lib version to lose the alpha
- merge all the PRs
- deploy

After deploy:
- go [here https://api.slack.com/apps](https://api.slack.com/apps)
- for all the 3 apps you see
- click on the app
- click on _Event Subscriptions_ in left sidebar
- scroll down to the section called _Subscribe to Workspace Events_
- [ ] make sure the _app_uninstalled_ event is there
- [ ] for staging and prod apps make sure _Enable Events_ is *On*

NB: if you have to add the _app_uninstalled_ event for our production app please check if a warning appears stating that we have to resubmit the app for approval. I don't think it will be required because this event doesn't require any new scope, in fact it doesn't require any scope at all.
